cloudevents/event_v02.go

```diff
@@ -0,0 +1,260 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cloudevents
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+)
+
+const (
+	// V02CloudEventsVersion is the version of the CloudEvents spec targeted
+	// by this library.
+	V02CloudEventsVersion = "0.2"
+
+	// required attributes
+	fieldSpecVersion = "SpecVersion"
+	fieldID          = "Id"
+	fieldType        = "Type"
+	fieldSource      = "Source"
+)
+
+// V02EventContext represents the non-data attributes of a CloudEvents v0.2
+// event.
+type V02EventContext struct {
+	// The version of the CloudEvents specification used by the event.
+	SpecVersion string `json:"specversion"`
+	// The type of the occurrence which has happened.
+	Type string `json:"type"`
+	// A URI describing the event producer.
+	Source string `json:"source"`
+	// ID of the event; must be non-empty and unique within the scope of the producer.
+	ID string `json:"id"`
+	// Timestamp when the event happened.
+	Time time.Time `json:"time,omitempty"`
+	// A link to the schema that the `data` attribute adheres to.
+	SchemaURL string `json:"schemaurl,omitempty"`
+	// A MIME (RFC2046) string describing the media type of `data`.
+	// TODO: Should an empty string assume `application/json`, `application/octet-stream`, or auto-detect the content?
+	ContentType string `json:"contenttype,omitempty"`
+	// Additional extension metadata beyond the base spec.
+	Extensions map[string]interface{} `json:"-,omitempty"`
+}
+
+// AsV01 implements the ContextTranslator interface.
+func (ec V02EventContext) AsV01() V01EventContext {
+	ret := V01EventContext{
+		CloudEventsVersion: V01CloudEventsVersion,
+		EventID:            ec.ID,
+		EventTime:          ec.Time,
+		EventType:          ec.Type,
+		SchemaURL:          ec.SchemaURL,
+		ContentType:        ec.ContentType,
+		Source:             ec.Source,
+		Extensions:         make(map[string]interface{}),
+	}
+	for k, v := range ec.Extensions {
+		// eventTypeVersion was retired in v0.2
+		if strings.EqualFold(k, "eventTypeVersion") {
+			etv, ok := v.(string)
+			if ok {
+				ret.EventTypeVersion = etv
+			}
+			continue
+		}
+		ret.Extensions[k] = v
+	}
+	return ret
+}
+
+// AsV02 implements the ContextTranslator interface.
+func (ec V02EventContext) AsV02() V02EventContext {
+	return ec
+}
+
+// AsHeaders implements the BinarySender interface.
+func (ec V02EventContext) AsHeaders() http.Header {
+	h := http.Header{}
+	h.Set("CE-SpecVersion", ec.SpecVersion)
```

Comment:
Any reason why those header names aren't constants? They seem to be shared throughout this change.

Reply:
Done.